client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx

```diff
@@ -127,21 +128,13 @@ export default function IssueLicenseForm( {
 
 			switch ( error.code ) {
 				case 'missing_valid_payment_method':
-					errorMessage = translate(
-						'We could not find a valid payment method.{{br/}} ' +
-							'{{a}}Try adding a new payment method{{/a}} or contact support.',
-						{
-							components: {
-								a: (
-									<a
-										href={
-											'/partner-portal/payment-methods/add?return=/partner-portal/issue-license'
-										}
-									/>
-								),
-								br: <br />,
+					page.redirect(
+						addQueryArgs(
+							{
+								return: addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) ),
```

Comment:
Wouldn't this cause an endless retry loop as the `product` query parameter gets added every time here?

Reply:
The way I see this happening:

1. Agency tries to issue license
2. Agency is missing a payment method which means we redirect to `payment-methods/add` with a `/issue-license` return URL
  * One of two things happens here: 1) we automatically return the agency to `/issue-license` if they have a payment method anyway or 2) we don't do anything and will wait for the user to submit a new payment method.
3. Returns back to `/issue-license` with the product we should automatically try to issue because of the `product` parameter
4. Is sent to the `/assign-license` page with the new product

@vitozev is it the third step that you see as a potential problem?
If we - for some reason - redirect agencies to `/add-payment-method` because the partner object in Calypso green says they have a payment method but the API returns `missing_valid_payment_method` and therefor we end up in a loop?

Reply:
The retry loop could happen if the local state and the API disagreed about whether there is a payment method **even after 3 retries**.
To avoid this, the redirect to add a payment method now only happens if the local state agrees a payment method is not available. If the local state says that a payment method is available but the API does not agree within 3 retries, the previous error message about not being able to find a valid payment method will be displayed.

Three retries should be enough time for Stripe to send their webhook and if it isn't then we can't issue the license. We can increase the number of retries or the exponential backoff interval to give Stripe more time though.

Reply:
I just went ahead and bumped the retries to 5 which should be plenty.

Reply:
We currently have a condition on `payment-methods/add` form success callback (after the user adds a credit card), that fires the action to fetch all credit cards which then fires another action that calculates if there's a valid payment method and sets it on the local state, that is useful for not depending on the Stripe webhook to turn that flag.

Reply:
Since `page.redirect()` or `page()` don't actually do a full refresh of the page, the local state is persisted and when the user lands back to `/issue-license` they already have the `has_valid_payment_method` flag set to `true`, not depending on the Stripe webhook, therefore no infinite loop should happen.